File: MEANSTRIC - Tricky Means/Tricky Means.py
# Project name : SPOJ: MEANSTRIC - Tricky Means
# Author       : Wojciech Raszka
# Date created : 2019-02-24
# Description  :
# Status       : Accepted (23289108)
# Comment      :

from math import log, exp
from functools import reduce

def arithmeticMean(An, x, n):
    return (An*n + x)/(n + 1)

def  pythagoreanMeans(Mn, x):
    global n
    n += 1
    x = float(x)
    if isinstance(Mn, str):
        Mn = float(Mn)
        A = arithmeticMean(Mn, x, n)
        lG = arithmeticMean(log(Mn), log(x), n)
        iA = arithmeticMean(1/Mn, 1/x, n)
    else:
        A = arithmeticMean(Mn[0], x, n)
        lG = arithmeticMean(Mn[1], log(x), n)
        iA = arithmeticMean(Mn[2], 1/x, n)

    return (A, lG, iA)


T = int(input())

for t in range(T):
    k = int(input())
    n = 0
    if k > 1:
        X = reduce(pythagoreanMeans, input().split())
        A = round(X[0], 9)
        G = round(exp(X[1]), 9)
        H = round(1/X[2], 9)
    else:
        A = float(input())
        G = A
        H = A
    print('%.9f %.9f %.9f' % (A, G, H))

File: SPOJ/MEANSTRIC - Tricky Means/Tricky Means.cpp
/*
* Project name : SPOJ: MEANSTRIC - Tricky Means
* Author       : Wojciech Raszka
* E-mail       : gitpistachio@gmail.com
* Date created : 2023-04-21
* Description  :
* Status       : Accepted (31240034)
* Tags         : c++, math, arithmetic mean, harmonic mean, geometric mean
* Comment      :
*/

#include <iostream>
#include <cmath>
#include <iomanip>

using namespace std;

double arithmeticMean(double A[], int n){
  double sum_of_elements = 0;
  
  for(int i = 0; i < n; i++){
  	sum_of_elements += A[i];
  }

  return sum_of_elements/n;
}

double geometricMean(double A[], int n){
  double product_of_elements = 1;
  
  for(int i = 0; i < n; i++){
  	product_of_elements *= A[i];
  }

  return pow(product_of_elements, 1./n);
}

double harmonicMean(double A[], int n){
  double sum_of_reciprocals = 0;
  
  for(int i = 0; i < n; i++){
  	sum_of_reciprocals += 1/A[i];
  }
  
  return n/sum_of_reciprocals;
}

int main() {
  int no_of_tests;
  int n;
  double am, ag, ah;
  double A[100];

  cout << fixed << setprecision(9);
  
  cin >> no_of_tests;
  
  for (int i = 0; i < no_of_tests; i++){
  	cin >> n;
  	
  	for (int j = 0; j < n; j++){
  		cin >> A[j];
  	}
  	
  	am = arithmeticMean(A, n);
  	ag = geometricMean(A, n);
  	ah = harmonicMean(A, n);
  	
  	
  	cout << am << ' ' << ag << ' ' << ah << endl;
  }
  
  return 0;
}